Add a debug assertion for subclasses of CanMakeCheckedPtr to be always heap allocated

rniwa opened the pull request:
#### e0db2dcca8c753a8a6d24cef28a2239ea6f7995b
<pre>
Add a debug assertion for subclasses of CanMakeCheckedPtr to be always heap allocated
<a href="https://bugs.webkit.org/show_bug.cgi?id=301458">https://bugs.webkit.org/show_bug.cgi?id=301458</a>

Reviewed by Geoffrey Garen.

This PR adds a debug assertion in the destructor of CanMakeCheckedPtrBase to assert that
the deletion is initiated with operator delete of the corresponding class.

Since there are quite a few cases where CheckedPtr capable type is either a member variable
of another class or it&apos;s allocated in stack, this PR also adds the mechanism to suppress
the newly added debug assertion in CanMakeCheckedPtrBase, and deploys it in places that
currently do not use operator delete for destruction.

Finally, this PR also converts a few objects to be heap-allocated on its own instead of as
a member variable of another type or in stack.

No new tests since there should be no behavioral differences.

Canonical link: <a href="https://commits.webkit.org/302184@main">https://commits.webkit.org/302184@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c4900f9532dd12b326442b6bd6e8c21e4d8d173

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128272 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/550 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39104 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135664 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79749 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/02bc3949-369c-432d-ac49-8adb1390dafb) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130144 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/484 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/424 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97638 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65543 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3fa45e35-e19d-4c33-b0a7-0bed2d300760) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131220 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/318 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114915 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78228 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/053d96b7-e3c5-478e-819c-a632e13c161f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/308 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33022 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78952 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120298 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108696 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138118 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126732 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/398 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/377 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106175 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/433 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111256 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105977 "Found 1 new API test failure: WebKitGTK/TestWebKitWebXR:/webkit/WebKitWebXR/permission-request (failure)") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/315 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29808 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52688 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20038 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/445 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63367 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159758 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/351 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39902 "Found 17 new JSC stress test failures: wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.default-wasm, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.wasm-bbq, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.wasm-bbq-no-consts, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.wasm-collect-continuously, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.wasm-eager, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.wasm-eager-jettison, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.wasm-no-cjit, wasm.yaml/wasm/stress/osr-entry-live-fpr.js.wasm-slow-memory, wasm.yaml/wasm/stress/osr-entry-live-stack.js.default-wasm ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/420 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/410 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/dom/Document.h:
```diff
@@ -444,6 +444,12 @@ class Document
     USING_CAN_MAKE_WEAKPTR(EventTarget);
     USING_CAN_MAKE_CHECKEDPTR(ScriptExecutionContext);
 
+    void setBegunCheckedPtrDeletion()
```
I'd suggest "DidBegin" here and elsewhere instead of "Begun". We usually use the "did" prefix in code to avoid generic past tense English verb conjugation, because it tends to be so irregular.

rniwa replied:
Sure, fixed.

geoffreygaren commented on Source/WTF/wtf/CheckedRef.h:
```diff
@@ -303,12 +311,22 @@ template <typename StorageType, typename PtrCounterType> class CanMakeCheckedPtr
 
-template<typename T, DefaultedOperatorEqual defaultedOperatorEqual = DefaultedOperatorEqual::No>
-class CanMakeCheckedPtr : public CanMakeCheckedPtrBase<SingleThreadIntegralWrapper<uint32_t>, uint32_t> {
+template<typename T, DefaultedOperatorEqual defaultedOperatorEqual = DefaultedOperatorEqual::No, CheckedPtrDeleteCheckException deleteException = CheckedPtrDeleteCheckException::No>
```
I think exceptions like this are more manageable when they're a separate traits class, but probably that change isn't worth it since this is all going to disappear so soon.

rniwa replied:
Ok, we can always change if it turns out to be difficult to get rid of them in near term.

geoffreygaren commented on Source/WTF/wtf/CheckedRef.h:
```diff
@@ -276,14 +276,22 @@ template<typename P> struct DefaultHash<CheckedRef<P>> : PtrHash<CheckedRef<P>>
 
+    ~CanMakeCheckedPtrBase()
+    {
+        ASSERT(m_hasBegunDeletion || deleteException == CheckedPtrDeleteCheckException::Yes);
```
Might as well ASSERT_WITH_SECURITY_IMPLICATION.

rniwa replied:
Done.

rniwa committed:
Add a debug assertion for subclasses of CanMakeCheckedPtr to be always heap allocated
https://bugs.webkit.org/show_bug.cgi?id=301458

Reviewed by Geoffrey Garen.

This PR adds a debug assertion in the destructor of CanMakeCheckedPtrBase to assert that
the deletion is initiated with operator delete of the corresponding class.

Since there are quite a few cases where CheckedPtr capable type is either a member variable
of another class or it's allocated in stack, this PR also adds the mechanism to suppress
the newly added debug assertion in CanMakeCheckedPtrBase, and deploys it in places that
currently do not use operator delete for destruction.

Finally, this PR also converts a few objects to be heap-allocated on its own instead of as
a member variable of another type or in stack.

No new tests since there should be no behavioral differences.

Canonical link: https://commits.webkit.org/302184@main
